### PR TITLE
Improve error handling

### PR DESCRIPTION
**Pull request by Kelvin-M**
Proposal Fix https://github.com/SubstraFoundation/substra-backend/issues/217

Some improvement to a better handling of those issues

### CODE REVIEW

File: backend/substrapp/tests/common.py [hunk 1]
```diff
@@ -293,6 +293,10 @@ def __init__(self, status, content):
         self.status_code = status
         self.content = content
 
+    def raise_for_status(self):
+        if self.status_code != 200:
+            raise Exception(f'HTTP Exception : {self.status_code}')
+
```
Comment (AurelienGasser) [inline backend/substrapp/tests/common.py:299]: I might be missing something, but doesn't that apply only to tests? I'm not seeing anything checking the response  code of actual requests.

Comment (Kelvin-M) [reply backend/substrapp/tests/common.py:299]: Old code ! I will remove it

File: backend/substrapp/tasks/tasks.py [hunk 2]
```diff
@@ -357,6 +361,13 @@ def prepare_data_sample(directory, tuple_):
     from substrapp.models import DataSample
     for data_sample_key in tuple_['dataset']['keys']:
         data_sample = DataSample.objects.get(pk=data_sample_key)
+
+        if not os.path.exists(data_sample.path) or not os.path.isdir(data_sample.path):
+            raise Exception(f'Data Sample ({data_sample.path}) is missing in local db')
+
+        if not os.listdir(data_sample.path):
+            raise Exception(f'Data Sample ({data_sample.path}) is empty in local db')
+
```
[the remark this answers is not in the document]
Comment (Kelvin-M) [reply backend/substrapp/tasks/tasks.py:370]: Not but the error here is more explicit I think :) (better twice than none)

**Comment by AurelienGasser** [comment]
@Kelvin-M This doesn't seem to fix https://github.com/SubstraFoundation/substra-backend/issues/222 as the PR description implies

File: backend/substrapp/tasks/tasks.py [hunk 3]
```diff
@@ -338,6 +338,10 @@ def prepare_opener(directory, tuple_):
 
     datamanager = DataManager.objects.get(pk=data_opener_hash)
 
+    # verify that local db opener file exists
+    if not os.path.exists(datamanager.data_opener.path) or not os.path.isfile(datamanager.data_opener.path):
+        raise Exception(f'DataOpener file ({datamanager.data_opener.path}) is missing in local db ')
```
Comment (AurelienGasser) [inline backend/substrapp/tasks/tasks.py:343]: I find the `in local db` part to be confusing: why "local db" and not "local filesystem"? 

I'm not sure what "local db" refers to. I would assume that the local db is postgres.

**Comment by Kelvin-M** [comment]
> @Kelvin-M This doesn't seem to fix #222 as the PR description implies

I don't think this issue was caused by a 403
As you can see here https://github.com/SubstraFoundation/substra-backend/blob/master/backend/substrapp/utils.py#L242:L266 , we handle http issues by raising a Node Error
Do you have a setup to reproduce it ?

File: backend/substrapp/tasks/tasks.py (hunk 3 above)
Comment (Kelvin-M) [reply backend/substrapp/tasks/tasks.py:343]: I will change it to local storage

**Comment by AurelienGasser** [comment]
@KelvinM

- Re: issue #222. You're right, it's not caused by a 403. I updated the issue. Could you please remove it from this PR's description?
- Re: naming: You updated one place but not the two others.

**Comment by Kelvin-M** [comment]
> @Kelvinm
> 
> * Re: issue #222. You're right, it's not caused by a 403. I updated the issue. Could you please remove it from this PR's description?
> * Re: naming: You updated one place but not the two others.

Done

**Comment by Kelvin-M** [comment]
@samlesu Can you review if this solution if good for your issue (#217)

**Comment by Kelvin-M** [comment]
> Thanks @Kelvin-M. This will help debugging when the "data volumes" are not available.
> 
> Using dedicated exceptions would have been better in my opinion. As it would be inconsistant with the existing code base, that's good for me :)

Yes, I think there is a potential improvement by reworking how exception are generated and raised in the backend.

**Pull request merged**